# EDA: Diagnosing Diabetes
In this dataset, EDA skills will be used to help inspect, clean, and validate the data.
Note: This dataset is from the National Institute of Diabetes and Digestive and Kidney Diseases.
https://www.kaggle.com/datasets/uciml/pima-indians-diabetes-database/

## Initial Inspection

In [1]:
#1 import library
import pandas as pd
import numpy as np

In [2]:
#2 preview using .head
diabetes_data = pd.read_csv("diabetes.csv")
print(diabetes_data.head())

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  


In [3]:
#3 no. of columns
print(len(diabetes_data.columns))

9


In [4]:
#4 no. of rows
print(len(diabetes_data))

768


## Further Inspection

In [5]:
#5 To check "null" value in columns
print(diabetes_data.info())
print("******")
print(diabetes_data.isnull().sum())
print("******")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB
None
******
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age               

In [6]:
#6 Further check missing data in columns
print(diabetes_data.describe())

       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  

If you take a look at the minimum values for the five columns as below, you’ll notice that they are all 0.
[Glucose, BloodPressure, SkinThickness, Insulin, BMI]

In addition to the 0 values that show up for the five columns above, there appear to be additional outliers, such as:
The maximum value of the Insulin column is 846, which is abnormally high.
The maximum value of the Pregnancies column is 17. While having 17 pregnancies is not impossible, this case might be something to look further into to determine its accuracy.

In [7]:
#9 to replace the instances of 0 with NaN
diabetes_data[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']] = \
diabetes_data[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']].replace(0,np.nan)

In [8]:
#10 check for missing (null) values
print(diabetes_data.isnull().sum())

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


In [9]:
#11 Print out all of the rows that contain missing (null) values
print(diabetes_data[diabetes_data.isnull().any(axis=1)])
# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.any.html
# https://stackoverflow.com/questions/14247586/how-to-select-rows-with-one-or-more-nulls-from-a-pandas-dataframe-without-listin

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
5              5    116.0           74.0            NaN      NaN  25.6   
7             10    115.0            NaN            NaN      NaN  35.3   
..           ...      ...            ...            ...      ...   ...   
761            9    170.0           74.0           31.0      NaN  44.0   
762            9     89.0           62.0            NaN      NaN  22.5   
764            2    122.0           70.0           27.0      NaN  36.8   
766            1    126.0           60.0            NaN      NaN  30.1   
767            1     93.0           70.0           31.0      NaN  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                  

One thing you might notice is that most rows with missing data have missing values in more than one column. In fact, every single row with at least one missing value also has a missing value in the insulin column. This is a clue as to why this data is missing! If patients did not have their insulin measured, why might they also not have had these other measurements taken?
Depending on how much data is missing, you might choose to remove specific rows or impute the missing values somehow.

In [10]:
#13 take a closer look at the data types of each column in diabetes_data.
print(diabetes_data.dtypes)

Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


There is no unexpected result unmatch to its data type.

In [11]:
#14 To check if there is error in the input in Column[Outcome]
print(diabetes_data.Outcome.unique())

[1 0]


If there is really an error found in this column - say a letter "O" is input instead of int[0], the replacement of "0" by 0 using the code below is to be performed.
diabetes_data['Outcome'] = diabetes_data['Outcome'].replace('O',0)
diabetes_data['Outcome'] = diabetes_data['Outcome'].astype("int")
print(diabetes_data.Outcome.unique())

In [12]:
#16 Use "value_counts()" to see the distribution in the selected columns with datatype(int64)
print(diabetes_data.Pregnancies.value_counts())

1     135
0     111
2     103
3      75
4      68
5      57
6      50
7      45
8      38
9      28
10     24
11     11
13     10
12      9
14      2
15      1
17      1
Name: Pregnancies, dtype: int64


In [16]:
int_column = ["Age" , "Outcome", "Pregnancies", "Glucose", "BloodPressure"]
for i in int_column:
    print(diabetes_data[i].value_counts())
    print("******")

22    72
21    63
25    48
24    46
23    38
28    35
26    33
27    32
29    29
31    24
41    22
30    21
37    19
42    18
33    17
38    16
36    16
32    16
45    15
34    14
46    13
43    13
40    13
39    12
35    10
50     8
51     8
52     8
44     8
58     7
47     6
54     6
49     5
48     5
57     5
53     5
60     5
66     4
63     4
62     4
55     4
67     3
56     3
59     3
65     3
69     2
61     2
72     1
81     1
64     1
70     1
68     1
Name: Age, dtype: int64
******
0    500
1    268
Name: Outcome, dtype: int64
******
1     135
0     111
2     103
3      75
4      68
5      57
6      50
7      45
8      38
9      28
10     24
11     11
13     10
12      9
14      2
15      1
17      1
Name: Pregnancies, dtype: int64
******
99.0     17
100.0    17
111.0    14
129.0    14
125.0    14
         ..
191.0     1
177.0     1
44.0      1
62.0      1
190.0     1
Name: Glucose, Length: 135, dtype: int64
******
70.0     57
74.0     52
78.0     45
68.0     45
72.0     44